vllm/executor/cpu_executor.py

```diff
@@ -0,0 +1,139 @@
+        model_config = CPUExecutor._verify_and_get_model_config(model_config)
+        cache_config = CPUExecutor._verify_and_get_cache_config(cache_config)
+
+        self.model_config = model_config
+        self.cache_config = cache_config
+        self.lora_config = lora_config
+        self.parallel_config = parallel_config
+        self.scheduler_config = scheduler_config
+        self.device_config = device_config
+
+        # Instantiate the worker and load the model to CPU.
+        self._init_worker()
+
+        self._init_cache()
+
+    def _init_worker(self):
+        from vllm.worker.cpu_worker import Worker
+
+        assert self.parallel_config.world_size == 1, (
+            "CPUExecutor only supports single CPU socket currently.")
+
+        distributed_init_method = get_distributed_init_method(
+            get_ip(), get_open_port())
+        self.driver_worker = Worker(
+            self.model_config,
+            self.parallel_config,
+            self.scheduler_config,
+            self.device_config,
+            local_rank=0,
+            rank=0,
+            distributed_init_method=distributed_init_method,
+            lora_config=self.lora_config,
+            kv_cache_dtype=self.cache_config.cache_dtype,
+            is_driver_worker=True,
+        )
+        self.driver_worker.init_device()
+        self.driver_worker.load_model()
+
+    def _init_cache(self) -> None:
+        num_cpu_blocks = self.driver_worker.get_cpu_cache_block_num(
+            block_size=self.cache_config.block_size,
+            cpu_swap_space=self.cache_config.swap_space_bytes,
+            cache_dtype=self.cache_config.cache_dtype,
+        )
+
+        logger.info(f"# CPU blocks: {num_cpu_blocks}, "
+                    f"# GPU blocks: {0}")
+        if num_cpu_blocks <= 0:
+            raise ValueError("No available memory for the cache blocks. "
+                             "Try increasing `swap_space` when "
+                             "initializing the engine.")
+
+        max_seq_len = self.cache_config.block_size * num_cpu_blocks
+        if self.model_config.max_model_len > max_seq_len:
+            raise ValueError(
+                f"The model's max seq len ({self.model_config.max_model_len}) "
+                "is larger than the maximum number of tokens that can be "
+                f"stored in KV cache ({max_seq_len}). Try increasing "
+                "`swap_space` or decreasing `max_model_len` when "
+                "initializing the engine.")
+
+        # To reuse the cache management procedure, use cpu cache as 'gpu cache'.
+        self.cache_config.num_gpu_blocks = num_cpu_blocks  # type: ignore
+        self.cache_config.num_cpu_blocks = 0  # type: ignore
+
+        # Initialize the cache.
+        self.driver_worker.init_cache_engine(cache_config=self.cache_config)
+
+    @staticmethod
+    def _verify_and_get_model_config(config: ModelConfig) -> ModelConfig:
```

Comment:
I think these functions are only used for ```CPUExecutor```, so placing them inside the ```CPUExecutor``` namespace will avoid exposing them to the outside directly.

Reply:
I think you don't have to be worried about namespace if the function is defined as this file's (this module's) private function since it will not be used elsewhere?

Reply:
Okay, moved them out.